File: shared/src/main/scala/org/pac4j/play/scala/deadbolt2/Pac4jHandlerCache.scala
package org.pac4j.play.scala.deadbolt2

import scala.collection.mutable.Map
import scala.concurrent.ExecutionContext
import be.objectify.deadbolt.scala.{DeadboltHandler, HandlerKey}
import be.objectify.deadbolt.scala.cache.HandlerCache
import javax.inject.{Inject, Singleton}
import org.pac4j.core.config.Config
import org.pac4j.core.context.session.SessionStore
import org.pac4j.core.util.CommonHelper

case class ClientsHandlerKey(clients: String) extends HandlerKey

/**
  * @author Zenkie Zhu
  * @since 4.1.0
  */
@Singleton
class Pac4jHandlerCache @Inject()(config: Config, sessionStore: SessionStore, roleHandler: Pac4jRoleHandler)(implicit ec: ExecutionContext) extends HandlerCache {

  private val handlers: Map[HandlerKey, DeadboltHandler] = Map()

  private val defaultHandlerKey = ClientsHandlerKey("defaultHandler")

  private val defaultHandler = new Pac4jHandler(config, null, sessionStore, roleHandler);

  handlers += (defaultHandlerKey -> defaultHandler)

  override def apply(): DeadboltHandler = defaultHandler

  override def apply(handlerKey: HandlerKey): DeadboltHandler = {
    handlers.get(handlerKey).
      getOrElse(getOrBuildHandler(handlerKey).get)
  }

  protected def getOrBuildHandler(handlerKey: HandlerKey): Option[DeadboltHandler] = synchronized {
    handlers.get(handlerKey).orElse {
      Some(handlerKey).
        filter(_.isInstanceOf[ClientsHandlerKey]).
        map(_.asInstanceOf[ClientsHandlerKey]).
        map(_.clients).
        map(new Pac4jHandler(config, _, sessionStore, roleHandler)).
        orElse(buildCustomHandler(handlerKey)).
        map { handler =>
          handlers += (handlerKey -> handler)
          handler
        }
    }
  }

  protected def buildCustomHandler(handlerKey: HandlerKey): Option[DeadboltHandler] = None

  override def toString() = CommonHelper.toNiceString(this.getClass(), "handlers", handlers, "config", config,
    "executionContext", ec, "playSessionStore", sessionStore)
}

File: shared/src/main/java/org/pac4j/play/deadbolt2/Pac4jHandlerCache.java
package org.pac4j.play.deadbolt2;

import be.objectify.deadbolt.java.DeadboltHandler;
import be.objectify.deadbolt.java.cache.HandlerCache;
import org.pac4j.core.config.Config;
import org.pac4j.core.context.session.SessionStore;
import org.pac4j.core.util.CommonHelper;
import play.libs.concurrent.HttpExecutionContext;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.HashMap;
import java.util.Map;

/**
 * This is a simple handler cache to store {@link Pac4jHandler}. The deadbolt key is the pac4j clients parameter,
 * the default key being a <code>null</code> clients parameter.
 *
 * @author Jerome Leleu
 * @since 2.6.0
 */
@Singleton
public class Pac4jHandlerCache implements HandlerCache {

    private final Map<String, DeadboltHandler> handlers = new HashMap<>();

    private Config config;

    private HttpExecutionContext httpExecutionContext;

    private SessionStore sessionStore;

    private DeadboltHandler defaultHandler;

    private final Pac4jRoleHandler roleHandler;

    @Inject
    public Pac4jHandlerCache(final Config config, final HttpExecutionContext httpExecutionContext, final SessionStore sessionStore, final Pac4jRoleHandler roleHandler) {
        this.config = config;
        this.httpExecutionContext = httpExecutionContext;
        this.sessionStore = sessionStore;
        this.roleHandler = roleHandler;
        defaultHandler = new Pac4jHandler(config, httpExecutionContext, null, sessionStore, roleHandler);
        handlers.put("defaultHandler", defaultHandler);
    }

    @Override
    public DeadboltHandler apply(final String clients) {
        DeadboltHandler handler = handlers.get(clients);
        if (handler == null) {
            handler = getAndBuildHandler(clients);
        }
        return handler;
    }

    protected synchronized DeadboltHandler getAndBuildHandler(final String clients) {
        DeadboltHandler handler = handlers.get(clients);
        if (handler == null) {
            handler = new Pac4jHandler(config, httpExecutionContext, clients, sessionStore, roleHandler);
        }
        return handler;
    }

    @Override
    public DeadboltHandler get() {
        return defaultHandler;
    }

    @Override
    public String toString() {
        return CommonHelper.toNiceString(this.getClass(), "handlers", handlers, "config", config,
                "httpExecutionContext", httpExecutionContext, "playSessionStore", sessionStore);
    }
}
